torch/lib/THD/CMakeLists.txt

```diff
@@ -140,6 +139,15 @@ ENDIF()
 INCLUDE_DIRECTORIES(${CMAKE_CURRENT_SOURCE_DIR})
 
 ADD_LIBRARY(THD STATIC ${all_cpp})
+
+ADD_DEPENDENCIES(THD caffe2)
+target_include_directories(THD PRIVATE
+  ${CMAKE_SOURCE_DIR}/aten/src/TH # provides "THAllocator.h" to THCGeneral.h
+  ${CMAKE_BINARY_DIR}/aten/src # provides "ATen/TypeExtendedInterface.h" to ATen.h
+  ${CMAKE_BINARY_DIR}/caffe2/aten/src # provides <TH/THGeneral.h> to THC.h
+  ${CMAKE_BINARY_DIR}/caffe2/aten/src/THC # provides "THCGeneral.h" to THC.h
+  )
```

Comment:
Mumble mumble modern style cmake should set these include dirs

Reply:
what Ed is referring to, and what I mentioned with my comments before is that it's the responsibility of `caffe2` cmake target to automatically include these directories when THD is being compiled, because you declared that THD depends on `caffe2` target via `ADD_DEPENDENCIES(THD caffe2)`

Reply:
Yes, exactly like Ed and Soumith mentioned. Also, looking at the folders, it seems that when we set the modern inclusion paths, we should use build interface and install interface separately, see this link for details:

https://cmake.org/cmake/help/v3.0/command/target_include_directories.html

Reply:
btw I've pulled this cleanup out as a separate task, since it's not a real blocker https://github.com/pytorch/pytorch/issues/12945